document/core/exec/instructions.rst

```diff
@@ -863,101 +824,111 @@ Memory Instructions
 
-5. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+7. Pop the value :math:`\I32.\CONST~n` from the stack.
 
-6. Pop the value :math:`\I32.\CONST~dst` from the stack.
+8. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-7. If :math:`cnt = 0`, then:
+9. Pop the value :math:`\I32.\CONST~s` from the stack.
 
-   a. Return.
+10. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
 
-8. If :math:`cnt = 1`, then:
+11. Pop the value :math:`\I32.\CONST~d` from the stack.
 
-   a. Push the value :math:`\I32.\CONST~dst` to the stack.
+12. If :math:`s + n` is larger than the length of :math:`\X{mem}.\MIDATA` or :math:`d + n` is larger than the length of :math:`\X{mem}.\MIDATA`, then:
 
-   b. Push the value :math:`\I32.\CONST~src` to the stack.
+    a. Trap.
 
-   c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
+13. If :math:`n = 0`, then:
 
-   d. Execute the instruction :math:`\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}`.
+   a. Return.
+
+14. If :math:`d \leq s`, then:
+
+   b. Push the value :math:`\I32.\CONST~d` to the stack.
+
+   c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
 
-   e. Return.
+   e. Push the value :math:`\I32.\CONST~s` to the stack.
 
-9. If :math:`dst <= src`, then:
+   f. Execute the instruction :math:`\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}`.
 
-   a. Push the value :math:`\I32.\CONST~dst` to the stack.
+   a. Assert: due to the earlier check against the memory size, :math:`d+1 < 2^{32}`.
 
-   b. Push the value :math:`\I32.\CONST~src` to the stack.
+   e. Push the value :math:`\I32.\CONST~(d+1)` to the stack.
 
-   c. Push the value :math:`\I32.\CONST~1` to the stack.
+   d. Assert: due to the earlier check against the memory size, :math:`s+1 < 2^{32}`.
 
-   d. Execute the instruction :math:`\MEMORYCOPY`.
+   f. Push the value :math:`\I32.\CONST~(s+1)` to the stack.
 
-   e. Push the value :math:`\vconst_{\I32}(dst+1)` to the stack.
+15. Else:
 
-   f. Push the value :math:`\vconst_{\I32}(src+1)` to the stack.
+   a. Assert: due to the earlier check against the memory size, :math:`d+n-1 < 2^{32}`.
 
-10. Else:
+   b. Push the value :math:`\I32.\CONST~(d+n-1)` to the stack.
 
-   a. Push the value :math:`\vconst_{\I32}(dst+cnt-1)` to the stack.
+   c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
 
-   b. Push the value :math:`\vconst_{\I32}(src+cnt-1)` to the stack.
+   d. Assert: due to the earlier check against the memory size, :math:`s+n-1 < 2^{32}`.
 
-   c. Push the value :math:`\I32.\CONST~1` to the stack.
+   e. Push the value :math:`\I32.\CONST~(s+n-1)` to the stack.
```

Comment:
Ditto the store stack argument ordering issue from above.

Reply:
Done.